accepted/future-releases/extension-types/feature-specification.md

```diff
@@ -0,0 +1,1131 @@
+<code>Name&lt;T<sub>1</sub>, .. T<sub>s</sub>&gt;(...)</code>
+or
+<code>Name&lt;T<sub>1</sub>, .. T<sub>s</sub>&gt;.name(...)</code>
+is used to invoke these constructors, and the type of such an expression is
+<code>Name&lt;T<sub>1</sub>, .. T<sub>s</sub>&gt;</code>.
+
+*In short, extension type constructors appear to be very similar to
+constructors in classes, and they correspond to the situation
+where the enclosing class has a single, non-late, final instance
+variable, which is initialized according to the normal rules for
+constructors (in particular, it can occur by means of `this.id`, or in
+an initializer list, or by an initializing expression in the
+declaration itself, but it is an error if it does not occur at all).*
+
+An extension type `V` used as an expression (*a type literal*) is allowed
+and has static type `Type`.
+
+An extension type declaration _DV_ can have the type modifier
+`final`. In this case it is a compile-time error for another extension type
+declaration _DV2_ to have _DV_ as a superinterface, if _DV2_ is declared in
+a different library.
+
+*As the grammar shows, any occurrence of the keywords `abstract`,
+`base`, `interface`, `sealed`, or `mixin` in an extension type
+declaration is a syntax error.*
+
+
+### Composing Extension Types
+
+This section describes the effect of including a clause derived from
+`<interfaces>` in an extension type declaration. We use the phrase
+_the implements clause_ to refer to this clause.
+
+*The rationale is that the set of members and member implementations of a
+given extension type may need to overlap with that of other extension type
+declarations. The implements clause allows for implementation reuse by
+putting some shared members in an extension type `V`, and including `V` in
+the implements clause of several extension type declarations
+<code>V<sub>1</sub> .. V<sub>k</sub></code>, thus "inheriting" the members
+of `V` into all of <code>V<sub>1</sub> .. V<sub>k</sub></code> without code
+duplication.*
+
+*The reason why this mechanism uses the keyword `implements` rather
+than `extends` to declare a relation that involves "inheritance" is
+that it has a similar semantics as that of extension members (in that
+they are statically resolved, and each member is applicable to every
+subtype).*
+
+Assume that _DV_ is an extension type declaration named `Name`, and
+`V1` occurs as one of the `<type>`s in the `<interfaces>` of _DV_. In
+this case we say that `V1` is a _superinterface_ of _DV_.
+
+If _DV_ does not include an `<interfaces>` clause then _DV_ has 
+`Object?` or `Object` as a direct superinterface, according to the subtype
+relation which was specified earlier.
+
+A compile-time error occurs if `V1` is a type name or a parameterized type
+which occurs as a superinterface in an extension type declaration _DV_, but
+`V1` does not denote an extension type, and `V1` does not denote a
+supertype of the ultimate representation type of _DV_.
```

Comment:
The `V1` is not a supertype, or the extension type erasure of `V1` is not a supertype of the extension type erasure of `DV`?

Reply:
But this would be what is written below: A compile-time error occurs if `R` is not a subtype of `S`. So, I don't understand the meaning of this paragraph.

Reply:
Hm... Does it mean maybe that we can write this?
```dart
class A {}

class B {}

extension type C(B it) implements A {}
```
But why?

Reply:
In the case where `V1` denotes an extension type, there's nothing new compared to the old inline class feature specification.

In the case where `V1` denotes a non-extension type, it is a situation where [this feature](https://github.com/dart-lang/language/pull/3138) is used. That feature was not part of the inline class proposal (when it landed), so that's a new thing.

With this, you can do exactly this kind of thing (note that we can only `implements` a supertype of the representation type):

```dart
class A {}

class B implements A {}

extension type C(B it) implements A {}
```

This allows the extension type to be "open" (to some extent, here we announce `A` even though we could have used `B` to make it even more open).

The point is that this kind of (open) extension type allows the representation type to work much like a "sticky" extension type:

First, we have access to all the members of `A`, so we can treat a `C` as if it had been an `A` (with respect to member invocations). Next, a `C` is assignable to a variable of type `A` (because of the subtype relationship, and similarly for a `List<C>` vs. a variable of type `List<A>`, etc.). This implies that the extension type works very much like a regular `extension` declaration. Another way to describe these behaviors is that an expression of type `C` makes it very easy to drop the extension type (that is, to _leave_ the abstraction). So you'd use it in cases where that is exactly what you want.

However, the extension members (that is, the members declared by `C` or "inherited" by `C` from some other extension type) are available on an expression of type `C` even in the case where the library that contains the member invocation does not directly import the library where `C` is declared. This makes the extension type different (and, in some situations, considerably more convenient) than the existing `extension` declarations.

Finally, again based on the subtype relationship, you can use promotion to _enter_ the abstraction very smoothly (again, you'd use this mechanism exactly when you want this): `List<A> list` can be promoted to `List<C>` (which requires a run-time type of `List<B>`, because the run-time value of `C` is `B`), and this means that each element of that list can be used as a `C` without further type tests or casts.